lib-js/test-case/test-case.ts

```diff
@@ -0,0 +1,249 @@
+
+  constructor(
+    dir: SpecDirectory,
+    impl: string,
+    compiler: Compiler,
+    todoMode?: string
+  ) {
+    this.dir = dir
+    this.impl = impl
+    this.compiler = compiler
+    this.todoMode = todoMode
+  }
+
+  /** Return whether this test directory has an indented sass file */
+  isIndented() {
+    return this.dir.hasFile("input.sass")
+  }
+
+  /** Return the name of the input file of this test directory. */
+  inputFile() {
+    return this.isIndented() ? "input.sass" : "input.scss"
+  }
+
+  /** Get the contents of the input file for this test directory. */
+  async input() {
+    const inputFile = this.isIndented() ? "input.sass" : "input.scss"
+    return await this.dir.readFile(inputFile)
+  }
+
+  /** Get the precision for this test directory */
+  async precision() {
+    return (await this.dir.options())[":precision"] ?? 10
+  }
+
+  expectsSuccess() {
+    return (
+      this.dir.hasFile(`output-${this.impl}.css`) ||
+      (this.dir.hasFile("output.css") &&
+        !this.dir.hasFile(`error-${this.impl}`))
+    )
+  }
+
+  private getResultFile(type: "output" | "warning" | "error") {
+    const ext = type === "output" ? ".css" : ""
+    const overrideFile = `${type}-${this.impl}${ext}`
+    return this.dir.hasFile(overrideFile) ? overrideFile : `${type}${ext}`
+  }
+
+  async expected(): Promise<SassResult> {
+    const isSuccessCase = this.expectsSuccess()
+    const resultFilename = this.getResultFile(
+      isSuccessCase ? "output" : "error"
+    )
+
+    let warning
+    // check if there's a warning
+    const warningFilename = this.getResultFile("warning")
+    if (this.dir.hasFile(warningFilename)) {
+      warning = await this.dir.readFile(warningFilename)
+    }
+
+    const expected = await this.dir.readFile(resultFilename)
+
+    if (isSuccessCase) {
+      return { isSuccess: true, output: expected, warning }
+    } else {
+      return { isSuccess: false, error: expected }
+    }
+  }
+
+  // Run the compiler and calculate the actual result
+  private async calcActualResult(): Promise<SassResult> {
+    const precision = await this.precision()
+
+    const cmdArgs = []
+    // Pass in the indentend option to the command
+    if (this.isIndented()) {
+      cmdArgs.push(this.impl === "dart-sass" ? "--indented" : "--sass")
+    }
+    if (precision) {
+      cmdArgs.push(`--precision`)
+      cmdArgs.push(`${precision}`)
+    }
+    cmdArgs.push(this.inputFile())
+
+    const { stdout, stderr, status } = await this.compiler.compile(
+      this.dir.path,
+      cmdArgs
+    )
+    if (status === 0) {
+      return { isSuccess: true, output: stdout, warning: stderr }
+    } else {
+      return { isSuccess: false, error: stderr }
+    }
+  }
+
+  async run(): Promise<TestResult> {
+    if (this._result) {
+      throw new Error(`Test case ${this.dir.relPath()} has already been run`)
+    }
+    // Remember to cache the results of the run, regardless of result
+    this._result = await this.doRun()
```

Comment:
Having the test case track both the intrinsic properties of the test itself as derived from the filesystem *and* the contingent properties that depend on how the user is interacting with it seems like a separation of concerns violation. I'd rather have the test runner keep a map from test cases to results.

Reply:
I think the problem with keeping the test cases and test results separate but linked with a mapping is that, in interactive mode, the user can *change* the test case through the various overwriting options. You have to pass in the test case anyways in order to overwrite it.

I don't think it's a separation of concerns violation because the test case and its results are linked: if you mark a test result as passing, you have to update the test contents so that it passes.

That said, I do think having to call `run()` is cumbersome and have some ideas on how to fix that (e.g. making sure the test results are calculated when constructing the TestCase object)